File: 35_search_insert_position/35_search_insert_position.py
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start = 0
        end = len(nums) - 1
        if end == -1:
            return 0
        while start + 1 < end:
            mid = start + (end - start) / 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                start = mid
            if nums[mid] > target:
                end = mid
        if nums[start] == target:
            return start
        elif nums[end] == target:
            return end
        elif nums[start] > target:
            return 0
        elif nums[end] < target:
            return len(nums)
        else:
            return start + 1

File: 35_search_insert_position/35_search_insert_position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        if (nums.size() == 0) {
            return NULL;
        }
        int ans = 0;
        int start = 0;
        int end = nums.size() - 1;
        while ((start + 1) < end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[mid] < target) {
                start = mid;
            }
            if (nums[mid] > target) {
                end = mid;
            }
        }
        if (nums[end] < target) {
            return end + 1;
        }
        if (nums[start] > target) {
            return (start > 0? start - 1 : 0);
        }
        if (nums[end] == target) {
            return end;
        }
        if (nums[start] == target) {
            return start;
        }
        return start + 1;
    }
};
